src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1407,26 +1407,41 @@ object Desugar {
                 } yield in.PredicateConstructor(proxy, idT, dArgs)(src)
 
-              case p@PMPredBase(_) =>
-                val recvTyp = info.typOfExprOrType(p.recv)
-                val recvT = typeD(recvTyp, Addressability.rValue)(src)
-                val argsTyp = info.typ(p.id) match {
-                  case FunctionT(args, AssertionT) => args
-                  case at: AbstractType =>
-                    Violation.violation(at.typing.isDefinedAt(Vector(recvTyp)), s"expected that built-in mpredicate ${p.recvWithId} is defined for receiver $recvT")
-                    at.typing(Vector(recvTyp)) match {
-                      case FunctionT(args, AssertionT) => args
-                      case t => violation(s"desugarer: $t cannot be converted to a predicate type")
-                    }
-                  case t => violation(s"desugarer: $t cannot be converted to a predicate type")
-                }
-                val argsT = argsTyp.map(typeD(_, Addressability.rValue)(src))
-                val proxy = getMPredProxy(recvT, argsT)
-                val idT = in.PredT(recvT +: argsT, Addressability.rValue)
-                for {
-                  dRecv <- exprAndTypeAsExpr(ctx)(p.recv)
-                  dArgs <- sequence(args.map { x => option(x.map(exprD(ctx)(_))) })
-                } yield in.PredicateConstructor(proxy, idT, Some(dRecv) +: dArgs)(src)
+              case p@PDottedBase(recvWithId) => info.resolve(recvWithId) match {
+                case Some(_: ap.ReceivedPredicate) =>
+                  val recvTyp = info.typOfExprOrType(p.recv)
+                  val recvT = typeD(recvTyp, Addressability.rValue)(src)
+                  val argsTyp = info.typ(p.id) match {
+                    case FunctionT(args, AssertionT) => args
+                    case at: AbstractType =>
+                      Violation.violation(at.typing.isDefinedAt(Vector(recvTyp)), s"expected that built-in mpredicate ${p.recvWithId} is defined for receiver $recvT")
+                      at.typing(Vector(recvTyp)) match {
+                        case FunctionT(args, AssertionT) => args
+                        case t => violation(s"desugarer: $t cannot be converted to a predicate type")
+                      }
+                    case t => violation(s"desugarer: $t cannot be converted to a predicate type")
+                  }
+                  val argsT = argsTyp.map(typeD(_, Addressability.rValue)(src))
+                  val proxy = getMPredProxy(recvT, argsT)
+                  val idT = in.PredT(recvT +: argsT, Addressability.rValue)
+                  for {
+                    dRecv <- exprAndTypeAsExpr(ctx)(p.recv)
+                    dArgs <- sequence(args.map { x => option(x.map(exprD(ctx)(_))) })
+                  } yield in.PredicateConstructor(proxy, idT, Some(dRecv) +: dArgs)(src)
+
+                case Some(_: ap.Predicate) =>
+                  val idT = info.typ(p.recvWithId) match {
```

Comment:
looks good. But maybe you could get rid of the code duplication by having a function `constructFPredConstructor` that takes `id` resp. `p.recvWithId` as argument